RFCs/0014-ToJsonString.md

```diff
@@ -0,0 +1,124 @@
+
+# Examples
+
+Here is an example of a resource containing a string-type property using a JSON formatted string.
+```
+"MySecret": {
+    "Type": "AWS::SecretsManager::Secret",
+    "Properties": {
+        "Name": "MySecret",
+        "SecretString": "{\"password\":\"Pa$$word\",\"secretToken\":\"123\"}"
+    }
+}
+```
+
+With the new `Fn::ToJsonString` intrinsic function we can simplify to the following:
+
+## JSON
+```
+"MySecret": {
+    "Type": "AWS::SecretsManager::Secret",
+    "Properties": {
+        "Name": "MySecret",
+        "SecretString": {
+            "Fn::ToJsonString": {
+                "password":"Pa$$word",
+                "secretToken": "123"
+            }
+        }
+    }
+}
+```
+
+## YAML
+```
+MySecret:
+    Type: AWS::SecretsManager::Secret
+    Properties:
+        Name: MySecret
+        SecretString:
+            Fn::ToJsonString:
+                password: "Pa$$word"
+                secretToken: "123"
+```
+
+### Short Form Syntax
+```
+MySecret:
+    Type: AWS::SecretsManager::Secret
+    Properties:
+        Name: MySecret
+        SecretString:
+            !ToJsonString:
+                password: "Pa$$word"
+                secretToken: "123"
+```
+
+# Motivation
+
+A CloudFormation user may want to use JSON strings as input to a resource property. For example, `AWS::CloudWatch::Dashboard` requires a JSON string for the `DashboardBody` attribute. A user may even want to use a JSON string as input to attributes where only a general string is required, such as in `SecretString` of `AWS::SecretsManager::Secret`.
+
+Today, this can only be accomplished through the use of external tools to convert object bodies to JSON strings, or use workarounds such as YAML multiline syntax. Having a feature to automate this transformation can provide various benefits, such as improving development workflow, improving code readability, or being able to utilize JSON/YAML syntax highlighting in text editors.
+
+Open Github request regarding this: https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/78
+
+# Details
+
+`Fn::ToJsonString` is an intrinsic function that takes in a template block as input and converts it into an escaped JSON string.
+
+* It will be restricted to only be used as the value of string-type resource properties.
+* Intrinsic functions (e.g. `Fn::If`, `Ref`) or pseudo parameters (e.g. `AWS::NoValue`) can be used within the input template block, with the limitation that references can not be made to resource properties. The input template block will be processed by the intrinsic functions and pseudo parameters before it is converted to a string (i.e. the resolved value of the intrinsic function is converted to the output string, and not the intrinsic function itself).
```

Comment:
> with the limitation that references can not be made to resource properties

Hmm, this may be too limiting. For example, it's common to need to use resource properties in CloudWatch dashboard definitions. Will discuss with you offline to see if we can remove this limitation.

Reply:
Discussed offline, we will remove this limitation.